Fix _split_iteration_ranges handling gt handling for two-dimensional tiling

laithsakka opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #170170

statically_known_gt(size, remaining) may return False for symbolic
expressions like 64*u0 vs u0, because both could be 0. Similarly for
backed expressions like s25*(((s70 - 5)//4)) - s25 and
 (s25*(((s70 - 5)//4)) - s25)*64.
 
 We want to assume tensor sizes are not 0 and pass the gt check we do the following:
 ```
if A//B = C and C >= 1
then A = B * C + R
and assuming A!=0
 A must be > B .
```
Note if we dont do this then sv.statically_known_multiple_of(remaining[i], expr) 
during the else branch will fail, for example u0 is not multiple of 64*u0.
```    
  if current_group < len(remaining):
      return_getters.append(
          operator.itemgetter(add_range(current_group, size))
                            
```



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

laithsakka committed:
Fix _split_iteration_ranges handling gt handling for Two-dimensional tiling

[ghstack-poisoned]

laithsakka committed:
Update on "Fix _split_iteration_ranges handling gt handling for Two-dimensional tiling"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

laithsakka committed:
Update on "Fix _split_iteration_ranges handling gt handling for Two-dimensional tiling"


statically_known_gt(size, remaining) may return False for symbolic
expressions like ``64*u0`` vs `u0`` because both could be 0. Similary for
backed expressions like ``s25*(((s70 - 5)//4)) - s25`` and
``s25*(((s70 - 5)//4)) - s25)*64``
                
To address this issue if size is a known multiple of remaining AND they're not
 statically known equal, then we assume size > remaining.
Note if we dont do this then sv.statically_known_multiple_of(remaining[i], expr) 
during the else branch will fail, for example u0 is not multiple of 64*u0.
```    
  if current_group < len(remaining):
      return_getters.append(
          operator.itemgetter(add_range(current_group, size))
                            
```



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

laithsakka committed:
Update on "Fix _split_iteration_ranges handling gt handling for Two-dimensional tiling"


statically_known_gt(size, remaining) may return False for symbolic
expressions like ``64*u0`` vs `u0`` because both could be 0. Similary for
backed expressions like ``s25*(((s70 - 5)//4)) - s25`` and
``s25*(((s70 - 5)//4)) - s25)*64``
                
To address this issue if size is a known multiple of remaining AND they're not
 statically known equal, then we assume size > remaining.
Note if we dont do this then sv.statically_known_multiple_of(remaining[i], expr) 
during the else branch will fail, for example u0 is not multiple of 64*u0.
```    
  if current_group < len(remaining):
      return_getters.append(
          operator.itemgetter(add_range(current_group, size))
                            
```



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

eellison reviewed (commented):
Can you add a representative test?

laithsakka commented on torch/_inductor/codegen/simd.py:
```diff
@@ -783,9 +783,25 @@ def getter(flat_vars: list[sympy.Expr]) -> sympy.Expr:
+                        sv.statically_known_multiple_of(size, remaining[current_group])
+                        and not sv.statically_known_equals(
+                            size, remaining[current_group]
+                        )
```
is it wrong correctness wise if they are same size? i thought its just a perf issue but i am not quite sure tbh not very familiar with this code.  

if this would be wrong can you or someone from inductor take over this change if this is wrong? since with out this Xl-formers is blocked to use torch2.10?

laithsakka committed:
Update on "Fix _split_iteration_ranges handling gt handling for two-dimensional tiling"


address https://github.com/pytorch/pytorch/issues/169952
statically_known_gt(size, remaining) may return False for symbolic
expressions like `64*u0` vs `u0` because both could be 0. Similarly for
backed expressions like ``s25*(((s70 - 5)//4)) - s25`` and
``s25*(((s70 - 5)//4)) - s25)*64``
                
To address this issue if `size` is a known multiple of `remaining` AND they're not statically known equal, 
then we assume size > remaining. 


Note if we dont do this then sv.statically_known_multiple_of(remaining[i], expr) 
during the else branch will fail, for example u0 is not multiple of 64*u0.
```    
  if current_group < len(remaining):
      return_getters.append(
          operator.itemgetter(add_range(current_group, size))
                            
```



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

laithsakka committed:
Update on "Fix _split_iteration_ranges handling gt handling for two-dimensional tiling"


address https://github.com/pytorch/pytorch/issues/169952
statically_known_gt(size, remaining) may return False for symbolic
expressions like `64*u0` vs `u0` because both could be 0. Similarly for
backed expressions like ``s25*(((s70 - 5)//4)) - s25`` and
``s25*(((s70 - 5)//4)) - s25)*64``
                
To address this issue if `size` is a known multiple of `remaining` AND they're not statically known equal, 
then we assume size > remaining. 


Note if we dont do this then sv.statically_known_multiple_of(remaining[i], expr) 
during the else branch will fail, for example u0 is not multiple of 64*u0.
```    
  if current_group < len(remaining):
      return_getters.append(
          operator.itemgetter(add_range(current_group, size))
                            
```



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

laithsakka committed:
Update on "Fix _split_iteration_ranges handling gt handling for two-dimensional tiling"


address https://github.com/pytorch/pytorch/issues/169952
statically_known_gt(size, remaining) may return False for symbolic
expressions like `64*u0` vs `u0` because both could be 0. Similarly for
backed expressions like ``s25*(((s70 - 5)//4)) - s25`` and
``s25*(((s70 - 5)//4)) - s25)*64``
                
To address this issue if `size` is a known multiple of `remaining` AND they're not statically known equal, 
then we assume size > remaining. 


Note if we dont do this then sv.statically_known_multiple_of(remaining[i], expr) 
during the else branch will fail, for example u0 is not multiple of 64*u0.
```    
  if current_group < len(remaining):
      return_getters.append(
          operator.itemgetter(add_range(current_group, size))
                            
```



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

laithsakka commented on torch/_inductor/codegen/simd.py:
```diff
@@ -783,9 +783,25 @@ def getter(flat_vars: list[sympy.Expr]) -> sympy.Expr:
+                # statically_known_gt(size, remaining) may return False for symbolic
+                # expressions like 64*u0 vs u0, because both could be 0. Similarly for
+                # backed expressions like s25*(((s70 - 5)//4)) - s25 and
+                # (s25*(((s70 - 5)//4)) - s25)*64.
```
oh wait actually i am wrong my bad 
A is multiple of D they could be the same but not 0. C, C*K
in that case would 2d tiling be problematic?

laithsakka committed:
Update on "Fix _split_iteration_ranges handling gt handling for two-dimensional tiling"


address https://github.com/pytorch/pytorch/issues/169952
statically_known_gt(size, remaining) may return False for symbolic
expressions like `64*u0` vs `u0` because both could be 0. Similarly for
backed expressions like ``s25*(((s70 - 5)//4)) - s25`` and
``s25*(((s70 - 5)//4)) - s25)*64``
                
To address this issue if `size` is a known multiple of `remaining` AND they're not statically known equal, 
then we assume size > remaining. 


Note if we dont do this then sv.statically_known_multiple_of(remaining[i], expr) 
during the else branch will fail, for example u0 is not multiple of 64*u0.
```    
  if current_group < len(remaining):
      return_getters.append(
          operator.itemgetter(add_range(current_group, size))
                            
```



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

laithsakka committed:
Update on "Fix _split_iteration_ranges handling gt handling for two-dimensional tiling"


statically_known_gt(size, remaining) may return False for symbolic
expressions like 64*u0 vs u0, because both could be 0. Similarly for
backed expressions like s25*(((s70 - 5)//4)) - s25 and
 (s25*(((s70 - 5)//4)) - s25)*64.
 
 We want to assume tensor sizes are not 0 and pass the gt check we do the following:
 ```
if A//B = C and C >= 1
then A = B * C + R
and assuming A!=0
 A must be > B .
```
Note if we dont do this then sv.statically_known_multiple_of(remaining[i], expr) 
during the else branch will fail, for example u0 is not multiple of 64*u0.
```    
  if current_group < len(remaining):
      return_getters.append(
          operator.itemgetter(add_range(current_group, size))
                            
```



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

laithsakka committed:
Update on "Fix _split_iteration_ranges handling gt handling for two-dimensional tiling"


statically_known_gt(size, remaining) may return False for symbolic
expressions like 64*u0 vs u0, because both could be 0. Similarly for
backed expressions like s25*(((s70 - 5)//4)) - s25 and
 (s25*(((s70 - 5)//4)) - s25)*64.
 
 We want to assume tensor sizes are not 0 and pass the gt check we do the following:
 ```
if A//B = C and C >= 1
then A = B * C + R
and assuming A!=0
 A must be > B .
```
Note if we dont do this then sv.statically_known_multiple_of(remaining[i], expr) 
during the else branch will fail, for example u0 is not multiple of 64*u0.
```    
  if current_group < len(remaining):
      return_getters.append(
          operator.itemgetter(add_range(current_group, size))
                            
```



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

Lucaskabela committed:
Fix linter for simd.py
